### PR TITLE
util: add vaultDestroyKeys option to destroy Vault kv-v2 secrets

### PR DESCRIPTION
**Pull request by nixpanic**
Hashicorp Vault does not completely remove the secrets in a kv-v2
backend when the keys are deleted. The metadata of the keys will be
kept, and it is possible to recover the contents of the keys afterwards.

With the new `vaultDestroyKeys` configuration parameter, this behaviour
can now be selected. By default the parameter will be set to `true`,
indicating that the keys and contents should completely be destroyed.
Setting it to any other value will make it possible to recover the
deleted keys.

See-also: https://github.com/libopenstorage/secrets/pull/55

---

<details>
<summary>Show available bot commands</summary>

These commands are normally not required, but in case of issues, leave any of
the following bot commands in an otherwise empty comment in this PR:

- `/retest ci/centos/<job-name>`: retest the `<job-name>` after unrelated
  failure (please report the failure too!)
- `/retest all`: run this in case the CentOS CI failed to start/report any test
  progress or results

</details>

### CODE REVIEW

**Comment by Madhu-1** [comment]
@nixpanic CI failure.

**Comment by Madhu-1** [comment]
> /test ci/centos/mini-e2e-helm/k8s-1.19

`X Exiting due to GUEST_PROVISION: Failed to start host: creating host: create: Error creating machine: Error in driver during machine creation: IP not available after waiting: machine minikube didn't return IP after 1 minute`

**Review by Madhu-1 — COMMENTED** [review]
LGTM. @nixpanic can we update E2E to verify keys are deleted completely?

**Comment by nixpanic** [comment]
> LGTM. @nixpanic can we update E2E to verify keys are deleted completely?

Sure, we default to kv-v2 in our testing, and that always supports the metadata for keys. A new commit has been added that enables checking for the metadata after deletion (should have been removed as well).

**Comment by nixpanic** [comment]
@mergifyio rebase

#2349 has been merged, pulling the CentOS images should work again.

File: internal/util/vault.go [hunk 1]
```diff
@@ -268,6 +292,25 @@ func (vc *vaultConnection) Destroy() {
...
+// `vaultDestroyKeys` configuration parameter.
+//
+// Setting the option `DestroySecret` option in the KeyContext while creating
+// new keys, causes failures, so the option only needs to be set in the
+// RemoveDEK() calls.
+func (vc *vaultConnection) getDeleteKeyContext() map[string]string {
+	keyContext := map[string]string{}
+	for k, v := range vc.keyContext {
+		keyContext[k] = v
+	}
```
Comment (Madhu-1) [inline internal/util/vault.go:312]: @nixpanic do we need this new map generation or add key to existing map?

**Review by Madhu-1 — DISMISSED** [review]

**Comment by Madhu-1** [comment]
> /test ci/centos/mini-e2e-helm/k8s-1.20

`    Aug  3 14:41:07.037: timed out waiting for the rbd-nbd process: rbd-nbd process is not running yet: command terminated with exit code 1`

**Comment by Madhu-1** [comment]
@pkalever Looks like some nbd related failed https://jenkins-ceph-csi.apps.ocp.ci.centos.org/blue/rest/organizations/jenkins/pipelines/mini-e2e-helm_k8s-1.20/runs/1891/nodes/110/steps/113/log/?start=0 PTAL

File: internal/util/vault.go (hunk 1 above)
Comment (nixpanic) [reply internal/util/vault.go:312]: Adding it to the existing map will cause failures during non-delete operations. So, creating a copy of the map and adding it there.

**Comment by nixpanic** [comment]
Rebasing with resolving merge conflicts was needed. Please review again, @Madhu-1 @humblec

**Review by Madhu-1 — APPROVED** [review]

File: go.mod [hunk 2]
```diff
@@ -25,7 +25,7 @@ require (
 	google.golang.org/grpc v1.38.0
 	k8s.io/api v0.21.2
 	k8s.io/apimachinery v0.21.2
-	k8s.io/client-go v0.21.2
+	k8s.io/client-go v12.0.0+incompatible
```
Comment (Rakshith-R) [inline go.mod:28]: I am not sure, but should we be concerned about this ? 
@Madhu-1 @nixpanic @humblec

Comment (nixpanic) [reply go.mod:28]: I do not think so, but we can add the v0.21.2 version in the `replace` section if needed.

Comment (Rakshith-R) [reply go.mod:28]: `libopenstorage/secrets` has a `replace` internally too, let's add `k8s.io/client-go => k8s.io/client-go v0.21.2` here too ?
https://github.com/libopenstorage/secrets/blob/dde442ea20ec9d59c71cea5ee0f21eeffd17ed19/go.mod#L38

Comment (nixpanic) [reply go.mod:28]: Just wanted to add this, but it already is there :rofl: See below on line 48.

Comment (Rakshith-R) [reply go.mod:28]: Copy that.😂🤦‍♂️

**Review by Rakshith-R — APPROVED** [review]

**Review by humblec — APPROVED** [review]

**Comment by nixpanic** [comment]
@mergifyio rebase

**Pull request merged**